src/jit/compiler.hpp

```diff
@@ -2708,6 +2708,26 @@ bool                    Compiler::compCanEncodePtrArgCntMax()
 #endif
 }
 
+// fgAreNextNodesValidForTailCall - Returns true if the fgMorphStmt contains the right pattern that 
+//                                  the tail call optimization expects.
+//
+// Arguments:
+//    fgMorphStmt      - The statement that contains the call expression to be converted to a tail call.
+inline
+bool Compiler::fgAreNextNodesValidForTailCall(GenTreePtr fgMorphStmt)
+{
+    GenTreePtr nextStmtExpr = fgMorphStmt->gtNext->gtStmt.gtStmtExpr;
+    return (nextStmtExpr->gtOper == GT_RETURN ||
+            // There could be a GT_COMMA generated between the call and the return.
+            // This happens when the result of the call is not used and discarded (in IL) by a pop.
+            // The generation of the pop IL instruction happens in "case CEE_POP:" in importer.cpp. The GT_COMMA
+            // is generated by the  gtUnusedValNode(op1) call and is intended to help the optimizers not to 
+            // change the call and its result in case of the call having side effects or the compilation happens
+            // for Debug configuration. Handle the potential existance of comma by checking that its second 
+            // operand is NOP and the next statement is a GT_RETURN.
+                (nextStmtExpr->gtOper == GT_COMMA && nextStmtExpr->gtOp.gtOp2->gtOper == GT_NOP &&
```

Comment:
I did not understand why importer for CEE_POP case is adding a GT_COMMA node.

Note that the IL is 

call
pop
ret

so when pop opcode is imported GT_CALL is on the top of the type stack. That is op1 = GT_CALL and considered to have a side-effect.  But the following if check is specifically avoiding adding a GT_COMMA node

```
                // If 'op1' is an expression, create an assignment node.
                // Helps analyses (like CSE) to work fine.

                if (op1->gtOper != GT_CALL)
                {
                    op1 = gtUnusedValNode(op1);
                }
```